Add explicit abort() at the end of AlignFail

aladram opened the pull request:
GCC cannot infer that `LogMessage::Finish()` (called indirectly by `GOOGLE_LOG(FAIL)` in `AlignFail`) exits program when `level_ == LOGLEVEL_FATAL`. Compilation will create a warning in projects importing the library for debug builds. The attribute cannot be individually assigned to `LogMessage::Finish()` because it does return conditionally of the log level. Tried with GCC 9.3.0 and GCC 12.1.0.

This change adds an explicit abort() for compiler to clearly understand this does not return, and enables `[[noreturn]]` for MSVC as well.

Closes #9758.

aladram commented:
If someone can take a look at this PR that would be awesome. Thanks!

aladram commented:
@fowles if you have some time to look at this that would be great! Fixes #9758.

fowles commented:
This particular code is very sensitive to performance on clang and I have a vague worry that we will get worse codegen as a result of this.

aladram commented:
Thank you for this quick answer! There are other solutions (remove only the attribute for GCC, insert a dummy assert, ...) but I believe this change does not impact performance for release code since this code is guarded by a `#ifndef NDEBUG` preprocessor directive.

aladram commented:
> It is currently guarded with that

I meant that, because of this guard, my change only impact debug builds, where performance should be less of a concern.

> I like the idea of adding some code that gcc will recognize as being no return

I can amend my PR to add something like this at the end of the function (extracted from `LogMessage::Finish`):
```cpp
void AlignFail(uintptr_t address) {
  GOOGLE_LOG(FATAL) << "Unaligned (" << align << ") access at " << address;

  // Explicit abort to tell compiler 
  abort();
}
```
I just want to point out that static code analyzers might flag this as dead code, so if you do use one we might need to add directives to ignore that. Please let me know what you think is best here :) and thanks again for your reactivity, appreciated!

fowles commented:
let's try adding the abort call and seeing what our CI thinks of it!

aladram commented:
I updated this PR to add the `abort()` we talked about, and dropped the exception for MSVC as well, since I suspect it was placed here to address the same issue. We can always restore it if we need to. Thanks!

aladram commented:
> just gave our CI system a kick. Will check back in 5 hours or so
Awesome thanks :)

aladram commented:
Seems like the C# CI job failed because of a package download issue (`Response status code does not indicate success: 429 (Too Many Requests).`), seems intermittent so maybe you can trigger it again?

Sil3ntStorm commented:
Curious, this dropped the `[[noexcept]]` for all compilers, despite also adding code that makes it explicit to every compiler that this function cannot ever return.
If there were concerns about code generation as a result of removing the noreturn attribute, won't those concerns still exist?
Either of the two changes would have been enough to fix the issue.

fowles commented:
Out thought was simply that `std::abort` does enough tha the compiler will figure out the rest on its own.  Happy to entertain PRs to get this more correct

Sil3ntStorm commented:
It was actually more of a curiosity question since when I looked to report this issue for clang, I found the issue I linked above and wanted to create a PR. However I saw the code in main already did both of the options I intended to put up for discussion in a PR, so I looked up the PR and read it.
I don't actually know enough about how compilers do code generation to have any input on whether or not it actually has any impact in any scenario.
However, this should cause clang to emit `-Wmissing-noreturn` now, so I guess I will be opening a PR shortly.
https://godbolt.org/z/ef9E1G9ah

aladram commented:
I confirm this is a mistake, I apparently removed noreturn but the goal was to add it without any compiler guard, thank you for noticing and for the PR!
